File: src/infra/config/manager.ts
import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';
import { z } from 'zod';
import { Config } from '../../domain/events/types';
import { DEFAULT_SECURITY_MASKS } from '../../shared/constants';

// Simple configuration schema - minimal options that users can modify
export const UserConfigSchema = z.object({
  enabled: z.boolean().default(true),
  language: z.enum(['ko', 'en']).default('ko'), 
  commitTrailer: z.boolean().default(true)
});

export type UserConfig = z.infer<typeof UserConfigSchema>;

export class ConfigManager {
  private static readonly CONFIG_FILE = '.sayu.yml';
  private userConfig: UserConfig;

  constructor(repoRoot: string) {
    this.userConfig = this.loadUserConfig(repoRoot);
  }

  private loadUserConfig(repoRoot: string): UserConfig {
    const configPath = path.join(repoRoot, ConfigManager.CONFIG_FILE);
    
    let rawConfig = {};
    
    if (fs.existsSync(configPath)) {
      try {
        const content = fs.readFileSync(configPath, 'utf-8');
        rawConfig = yaml.load(content) || {};
      } catch (error) {
        console.warn(`Failed to load config:`, error);
      }
    }
    
    return UserConfigSchema.parse(rawConfig);
  }

  // Apply user settings + environment variable overrides
  getUserConfig(): UserConfig {
    const config = { ...this.userConfig };
    
    // Environment variable overrides
    if (process.env.SAYU_ENABLED === 'false') config.enabled = false;
    if (process.env.SAYU_LANG === 'en' || process.env.SAYU_LANG === 'ko') {
      config.language = process.env.SAYU_LANG;
    }
    if (process.env.SAYU_TRAILER === 'false') config.commitTrailer = false;
    
    return config;
  }

  // Return full Config for backward compatibility (mostly fixed default values)
  get(): Config {
    const userConfig = this.getUserConfig();
    
    return {
      connectors: {
        claude: true,
        cursor: true, 
        editor: true,
        cli: { mode: 'zsh-preexec' },
        browser: { mode: 'off' }
      },
      window: { 
        beforeCommitHours: 24 
      },
      filter: {
        domainAllowlist: ['github.com', 'developer.mozilla.org', 'stackoverflow.com'],
        noise: { graceMinutes: 5, minScore: 0.6 }
      },
      summarizer: { 
        mode: 'hybrid', 
        maxLines: { commit: 12 } 
      },
      privacy: { 
        maskSecrets: true,
        masks: [...DEFAULT_SECURITY_MASKS] 
      },
      output: { 
        commitTrailer: userConfig.commitTrailer 
      }
    };
  }

  // SimpleConfig compatibility method
  getEffectiveConfig(): UserConfig {
    return this.getUserConfig();
  }

  save(repoRoot: string, newConfig: Partial<UserConfig>): void {
    this.userConfig = UserConfigSchema.parse({ ...this.userConfig, ...newConfig });
    
    const configPath = path.join(repoRoot, ConfigManager.CONFIG_FILE);
    const content = yaml.dump(this.userConfig, { 
      indent: 2,
      lineWidth: 120 
    });
    
    fs.writeFileSync(configPath, content, 'utf-8');
  }

  static createDefault(repoRoot: string): void {
    const configPath = path.join(repoRoot, ConfigManager.CONFIG_FILE);
    
    if (fs.existsSync(configPath)) {
      return;
    }

    const defaultConfig = `# Sayu Configuration
# AI automatically collects your development context

# Enable Sayu (set to false to disable)
enabled: true

# Language setting (ko: Korean, en: English)  
language: ko

# Add AI context to commit messages
commitTrailer: true

# Can also be configured via environment variables:
# SAYU_ENABLED=false
# SAYU_LANG=en  
# SAYU_TRAILER=false
#
# LLM API keys in .env file:
# GEMINI_API_KEY=your-key
# OPENAI_API_KEY=your-key
# ANTHROPIC_API_KEY=your-key
`;

    fs.writeFileSync(configPath, defaultConfig, 'utf-8');
    console.log(`Created config at ${configPath}`);
  }
}

// Alias for SimpleConfigManager replacement
export { ConfigManager as SimpleConfigManager };
export type { UserConfig as SimpleConfigType };

File: infra/config/manager.py
"""Configuration management for Sayu"""

import os
from pathlib import Path
from typing import Optional, Dict, Any
import yaml

from domain.events.types import Config, UserConfig
from shared import constants


class ConfigManager:
    """Manage Sayu configuration"""
    
    CONFIG_FILE = '.sayu.yml'
    
    def __init__(self, repo_root: str):
        """Initialize config manager for repository"""
        self.repo_root = Path(repo_root)
        self.user_config = self._load_user_config()
    
    def _load_user_config(self) -> UserConfig:
        """Load user configuration from file"""
        config_path = self.repo_root / self.CONFIG_FILE
        
        raw_config = {}
        if config_path.exists():
            try:
                with open(config_path, 'r', encoding='utf-8') as f:
                    raw_config = yaml.safe_load(f) or {}
            except Exception as e:
                print(f"Warning: Failed to load config: {e}")
        
        # Parse with defaults
        return UserConfig(
            enabled=raw_config.get('enabled', True),
            language=raw_config.get('language', 'ko'),
            commitTrailer=raw_config.get('commitTrailer', True)
        )
    
    def get_user_config(self) -> UserConfig:
        """Get user config with environment overrides"""
        config = UserConfig(
            enabled=self.user_config.enabled,
            language=self.user_config.language,
            commitTrailer=self.user_config.commitTrailer
        )
        
        # System environment variable overrides (always take precedence)
        if os.getenv('SAYU_ENABLED') == 'false':
            config.enabled = False
        if os.getenv('SAYU_LANG') in ['en', 'ko']:
            config.language = os.getenv('SAYU_LANG')
        if os.getenv('SAYU_TRAILER') == 'false':
            config.commitTrailer = False
        
        return config
    
    def get(self) -> Config:
        """Get full configuration for backward compatibility"""
        user_config = self.get_user_config()
        
        # Load full config from .sayu.yml if exists
        config_path = self.repo_root / self.CONFIG_FILE
        raw_config = {}
        if config_path.exists():
            try:
                with open(config_path, 'r', encoding='utf-8') as f:
                    raw_config = yaml.safe_load(f) or {}
            except Exception:
                pass
        
        return Config(
            connectors=raw_config.get('connectors', {
                'claude': True,
                'cursor': True,
                'editor': True,
                'cli': {'mode': 'zsh-preexec'},
                'git': True
            }),
            privacy=raw_config.get('privacy', {
                'maskSecrets': False,
                'masks': list(constants.DEFAULT_SECURITY_MASKS)
            }),
            output=raw_config.get('output', {
                'commitTrailer': user_config.commitTrailer
            })
        )
    
    def get_effective_config(self) -> UserConfig:
        """Get effective user configuration"""
        return self.get_user_config()
    
    def get_constants(self) -> Dict[str, Any]:
        """Get constants - no longer overridable via config"""
        # Return hardcoded constants
        return {
            'DEFAULT_LOOKBACK_HOURS': constants.DEFAULT_LOOKBACK_HOURS,
            'MAX_COMMIT_TRAILER_LINES': constants.MAX_COMMIT_TRAILER_LINES,
            'CACHE_TTL_SECONDS': constants.CACHE_TTL_SECONDS,
            'COLLECTOR_TIMEOUT_MS': constants.COLLECTOR_TIMEOUT_MS,
            'MAX_CONVERSATION_COUNT': constants.MAX_CONVERSATION_COUNT,
            'MAX_CONVERSATION_LENGTH': constants.MAX_CONVERSATION_LENGTH,
            'MAX_SIMPLIFIED_CONVERSATIONS': constants.MAX_SIMPLIFIED_CONVERSATIONS,
            'MAX_SIMPLIFIED_LENGTH': constants.MAX_SIMPLIFIED_LENGTH,
            'MAX_HIGH_VALUE_EVENTS': constants.MAX_HIGH_VALUE_EVENTS,
            'MIN_RESPONSE_LENGTH': constants.MIN_RESPONSE_LENGTH,
            'MAX_RAW_RESPONSE_LENGTH': constants.MAX_RAW_RESPONSE_LENGTH,
            'MAX_FILE_DISPLAY': constants.MAX_FILE_DISPLAY,
            'MAX_LINE_LENGTH': constants.MAX_LINE_LENGTH,
            'LLM_TEMPERATURE': constants.LLM_TEMPERATURE,
            'LLM_MAX_OUTPUT_TOKENS': constants.LLM_MAX_OUTPUT_TOKENS,
            'SUMMARY_SEPARATOR': constants.SUMMARY_SEPARATOR,
            'SUMMARY_FOOTER': constants.SUMMARY_FOOTER
        }
    
    def save(self, new_config: Dict[str, Any]):
        """Save updated configuration"""
        # Merge with existing
        merged = {
            'enabled': new_config.get('enabled', self.user_config.enabled),
            'language': new_config.get('language', self.user_config.language),
            'commitTrailer': new_config.get('commitTrailer', self.user_config.commitTrailer)
        }
        
        self.user_config = UserConfig(**merged)
        
        # Write to file
        config_path = self.repo_root / self.CONFIG_FILE
        with open(config_path, 'w', encoding='utf-8') as f:
            yaml.dump(merged, f, default_flow_style=False, sort_keys=False)
    
    @classmethod
    def create_default(cls, repo_root: str):
        """Create default configuration file"""
        config_path = Path(repo_root) / cls.CONFIG_FILE
        
        if config_path.exists():
            return
        
        default_content = """# Sayu Configuration
# 커밋에 '왜'를 남기는 개인 로컬 블랙박스

connectors:
  claude: true
  cursor: true
  editor: true
  cli:
    mode: "zsh-preexec"   # or "atuin" | "off"

privacy:
  maskSecrets: true       # 민감정보 마스킹 여부
  masks:                  # 추가 마스킹 패턴 (정규식)
    - "AKIA[0-9A-Z]{16}"  # AWS Access Key
    - "(?i)authorization:\\s*Bearer\\s+[A-Za-z0-9._-]+"

output:
  commitTrailer: true     # 커밋 메시지에 트레일러 추가

# 환경 변수로도 설정 가능:
# SAYU_ENABLED=false
# SAYU_LANG=en
# SAYU_TRAILER=false
"""
        
        with open(config_path, 'w', encoding='utf-8') as f:
            f.write(default_content)
        
        print(f"Created config at {config_path}")


# Alias for compatibility
SimpleConfigManager = ConfigManager
